contracts/v0.6/LinkTokenWrapper.sol

```diff
@@ -0,0 +1,94 @@
+
+// SPDX-License-Identifier: MIT
+pragma solidity >0.6.0 <0.8.0;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "./swap/ISwap.sol";
+import "./LinkToken.sol";
+
+/**
+ * @title LinkTokenWrapper
+ * @notice Using this contract users can lock source tokens to be minted wrap tokens.
+ * @dev Useful to add ERC677 functionality to ERC20 LINK bridge tokens.
+ */
+contract LinkTokenWrapper is ISwap, LinkToken {
+  using SafeERC20 for IERC20;
+
+  IERC20 private immutable _source;
+
+  /**
+   * @dev Create the LinkTokenWrapper token contract
+   *
+   * Requirements:
+   * - `sourceAddr` cannot be the zero address.
+   *
+   * @param sourceAddr The address of source token contract
+   */
+  constructor(address sourceAddr)
+    public
+  {
+    require(sourceAddr != address(0), "LinkTokenWrapper: sourceAddr is zero address");
+    _source = IERC20(sourceAddr);
+  }
+
+  /**
+   * @dev Overrides parent contract so no tokens are minted on deployment.
+   * @inheritdoc LinkToken
+   */
+  function _onCreate()
+    internal
+    override
+  {
+    // noop
+  }
+
+  /// @inheritdoc ISwap
+  function source()
+    external
+    view
+    override
+    returns (address)
+  {
+    return address(_source);
+  }
+
+  /// @inheritdoc ISwap
+  function target()
+    external
+    view
+    override
+    returns (address)
+  {
+    return address(this);
+  }
+
+  /**
+   * @dev Contract mints the `amount` of LinkTokenWrapper tokens.
+   *
+   * Requirements:
+   * - the contract must have allowance for `caller`'s source tokens of at least `amount`.
+   *
+   * @inheritdoc ISwap
+   */
+  function deposit(uint256 amount)
+    external
+    override
+  {
+    _source.safeTransferFrom(msg.sender, address(this), amount);
```

Comment:
where would we use this? we're going to use the peg, right?

Reply:
For future situations of expansion to network B, where they already deployed an ERC20 LINK that is a pain to upgrade...

In that case, instead of deploying LinkToken + Peg, we have an option to deploy just the LinkTokenWrapper.

Reply:
So in the event where a new chain deploys an ERC20 LINK token, instead of deploying our own separate ERC677 token, we can deploy this wrapper contract (which inherits `transferAndCall`) as our LINK token?

It looks as though holders of the ERC20 version would need to perform the extra step of calling `deposit` on the wrapper to convert their ERC20 to ERC677 LINK, right? (and consequently, `withdraw` to convert back)

Reply:
Yes, exactly! Bridge ERC20 LINK is the value bearing token and we just wrap it with additional functionality. Similar to how wETH adds ERC20 functionality to ETH.

We can also add additional functionality to skip some steps like `depositTo`, `withdrawTo`, `withdrawFrom` and also ERC677 like `depositAndCall` using the new [WETH10](https://github.com/WETH10/WETH10/blob/main/contracts/interfaces/IWETH10.sol) as inspiration.

Reply:
I really like this.